packages/store/addon/-private/system/request-cache.ts

```diff
@@ -0,0 +1,133 @@
+import { RecordIdentifier } from "./record-identifier";
+import { Request, FindRecordQuery, RequestState } from './fetch-manager';
+
+import {
+  _findHasMany,
+  _findBelongsTo,
+  _findAll,
+  _query,
+  _queryRecord,
+} from './store/finders';
+
+const touching = Symbol('touching');
+
+enum RequestStateEnum {
+  pending = 'pending',
+  fulfilled = 'fulfilled',
+  rejected = 'rejected'
+}
+
+export interface InternalRequest extends RequestState {
+  [touching]: RecordIdentifier[]
+}
+
+export default class RequestCache {
+
+  _pending: { [lid: string]: InternalRequest[] }
+  _done: { [lid: string]: InternalRequest[] }
+  _subscriptions: { [lid: string]: Function[] }
+
+  constructor() {
+    this._pending = Object.create(null);
+    this._done = Object.create(null);
+    this._subscriptions = Object.create(null);
+  }
+
+  enqueue(promise: Promise<any>, queryRequest: Request) {
+    if ('recordIdentifier' in queryRequest.data[0]) {
+      let query: FindRecordQuery = queryRequest.data[0];
+      let lid = query.recordIdentifier.lid;
+      let type = query.op === 'saveRecord' ? <const> 'mutation' : <const> 'query';
```

Comment:
```suggestion
      let type = query.op === 'saveRecord' ? 'mutation' as const : 'query' as const;
```
it's considered stylistically better to cast with `as` because of avoidance of JSX collision